go/arrow/compute/exprs/exec.go

```diff
@@ -0,0 +1,631 @@
+		defer arr.Release()
+		return compute.NewDatum(scalar.NewListScalar(arr)), nil
+	case *expr.MapLiteral:
+		dt, _, err := FromSubstraitType(v.Type, ext)
+		if err != nil {
+			return nil, err
+		}
+
+		mapType, ok := dt.(*arrow.MapType)
+		if !ok {
+			return nil, fmt.Errorf("%w: map literal with non-map type", arrow.ErrInvalid)
+		}
+
+		keys, values := make([]scalar.Scalar, len(v.Value)), make([]scalar.Scalar, len(v.Value))
+		for i, kv := range v.Value {
+			k, err := literalToDatum(mem, kv.Key, ext)
+			if err != nil {
+				return nil, err
+			}
+			defer k.Release()
+			scalarKey := k.(*compute.ScalarDatum).Value
+
+			v, err := literalToDatum(mem, kv.Value, ext)
+			if err != nil {
+				return nil, err
+			}
+			defer v.Release()
+			scalarValue := v.(*compute.ScalarDatum).Value
+
+			if !arrow.TypeEqual(mapType.KeyType(), scalarKey.DataType()) {
+				return nil, fmt.Errorf("%w: key type mismatch for %s, got key with type %s",
+					arrow.ErrInvalid, mapType, scalarKey.DataType())
+			}
+			if !arrow.TypeEqual(mapType.ValueType(), scalarValue.DataType()) {
+				return nil, fmt.Errorf("%w: value type mismatch for %s, got key with type %s",
+					arrow.ErrInvalid, mapType, scalarValue.DataType())
+			}
+
+			keys[i], values[i] = scalarKey, scalarValue
+		}
+
+		keyBldr, valBldr := array.NewBuilder(mem, mapType.KeyType()), array.NewBuilder(mem, mapType.ValueType())
+		defer keyBldr.Release()
+		defer valBldr.Release()
+
+		if err := scalar.AppendSlice(keyBldr, keys); err != nil {
+			return nil, err
+		}
+		if err := scalar.AppendSlice(valBldr, values); err != nil {
+			return nil, err
+		}
+
+		keyArr, valArr := keyBldr.NewArray(), valBldr.NewArray()
+		defer keyArr.Release()
+		defer valArr.Release()
+
+		kvArr, err := array.NewStructArray([]arrow.Array{keyArr, valArr}, []string{"key", "value"})
+		if err != nil {
+			return nil, err
+		}
+		defer kvArr.Release()
+
+		return compute.NewDatumWithoutOwning(scalar.NewMapScalar(kvArr)), nil
+	case *expr.StructLiteral:
+		fields := make([]scalar.Scalar, len(v.Value))
+		names := make([]string, len(v.Value))
+
+		s, err := scalar.NewStructScalarWithNames(fields, names)
+		return compute.NewDatum(s), err
+	case *expr.ProtoLiteral:
+		switch v := v.Value.(type) {
+		case *types.Decimal:
+			if len(v.Value) != arrow.Decimal128SizeBytes {
+				return nil, fmt.Errorf("%w: decimal literal had %d bytes (expected %d)",
+					arrow.ErrInvalid, len(v.Value), arrow.Decimal128SizeBytes)
+			}
+
+			var val decimal128.Num
+			data := (*(*[arrow.Decimal128SizeBytes]byte)(unsafe.Pointer(&val)))[:]
+			copy(data, v.Value)
+			if endian.IsBigEndian {
+				// reverse the bytes
+				for i := len(data)/2 - 1; i >= 0; i-- {
+					opp := len(data) - 1 - i
+					data[i], data[opp] = data[opp], data[i]
+				}
+			}
+
+			return compute.NewDatum(scalar.NewDecimal128Scalar(val,
+				&arrow.Decimal128Type{Precision: v.Precision, Scale: v.Scale})), nil
+		case *types.UserDefinedLiteral: // not yet implemented
```

Comment:
Return an error here?

Reply:
it'll hit the case of returning `arrow.ErrNotImplemented` which is at the bottom of the function for anything which doesn't return during the switch.